### ISSUE TITLE
[native-image] Micronaut + Hibernate results in Error encountered while analysing java.lang.reflect.Method.getDefaultValue() 

### ISSUE THREAD

**Issue by graemerocher** [oracle/graal#1295]
The application at https://github.com/graemerocher/micronaut-petclinic

Uses Micronaut, Hibernate and Postgres and works on GraalVM substrate rc16. Switching to the `graal19` branch which has the following changes:

https://github.com/graemerocher/micronaut-petclinic/commit/d3fde7574162eed04c0bf49309c8baab4a91db8c

The result is the following error:

```
[petclinic:84826]     analysis:  22,902.77 ms
Error: Field java.lang.reflect.Method.defaultValue is not present on type java.lang.reflect.Constructor. Error encountered while analysing java.lang.reflect.Method.getDefaultValue() 
Parsing context:
	parsing io.micronaut.inject.annotation.AnnotationMetadataSupport$AnnotationProxyHandler.invoke(AnnotationMetadataSupport.java:321)
	parsing com.sun.proxy.$Proxy257.hashCode(Unknown Source)
	parsing java.util.concurrent.CopyOnWriteArrayList.hashCode(CopyOnWriteArrayList.java:1062)
	parsing java.util.AbstractList.hashCode(AbstractList.java:541)
	parsing org.graalvm.collections.Equivalence$2.hashCode(Equivalence.java:86)
	parsing org.graalvm.collections.EconomicMapImpl.getHashIndex(EconomicMapImpl.java:411)
	parsing org.graalvm.collections.EconomicMapImpl.findHash(EconomicMapImpl.java:285)
	parsing org.graalvm.collections.EconomicMapImpl.find(EconomicMapImpl.java:254)
	parsing org.graalvm.collections.EconomicMapImpl.put(EconomicMapImpl.java:426)
	parsing org.graalvm.collections.EconomicMap.putAll(EconomicMap.java:84)
	parsing org.graalvm.collections.EconomicMapImpl.<init>(EconomicMapImpl.java:183)
	parsing org.graalvm.collections.EconomicMapImpl.create(EconomicMapImpl.java:158)
	parsing org.graalvm.collections.EconomicMap.create(EconomicMap.java:171)
	parsing org.graalvm.compiler.options.ModifiableOptionValues.update(ModifiableOptionValues.java:93)
	parsing com.oracle.svm.core.option.RuntimeOptionParser.parse(RuntimeOptionParser.java:164)

```

So the question is what combination of flags is required to get passed this and obtain the same behaviour that worked in rc16?

**Comment by graemerocher** [comment]
Also note that I don't currently understand why Substrate is arriving to `java.lang.reflect.Method.defaultValue` since there are no direct references in the code `io.micronaut.inject.annotation.AnnotationMetadataSupport$AnnotationProxyHandler.invoke`

**Comment by graemerocher** [comment]
Possibly I forgot to post the steps to reproduce (using 19):

```
1. git clone https://github.com/graemerocher/micronaut-petclinic.git
2. git checkout -b graal19
3. cd micronaut-petclinic
4. ./mvnw package
5. native-image -cp target/micronaut-petclinic-0.1.0.BUILD-SNAPSHOT.jar 
```

**Comment by ilopmar** [comment]
Some more information based on Graeme's repo. 

I've tried with Graal from master branch and the native image can be created after adding `org.apache.xerces.jaxp.validation.SimpleXMLSchema` to `--initialize-at-build-time`.

Then, when accessing the database the application fails with this error:
```
java.lang.NoClassDefFoundError: Could not initialize class org.hibernate.hql.internal.ast.HqlParser
at com.oracle.svm.core.hub.ClassInitializationInfo.initialize(ClassInitializationInfo.java:214)
at java.lang.Class.ensureInitialized(DynamicHub.java:451)
at org.hibernate.hql.internal.ast.QueryTranslatorImpl.parse(QueryTranslatorImpl.java:285)
at org.hibernate.hql.internal.ast.QueryTranslatorImpl.doCompile(QueryTranslatorImpl.java:189
```
So, I added the class `org.hibernate.hql.internal.ast.HqlParser` (and also `antlr.collections.impl.BitSet` because it failed again) to `--initialize-at-build-time` and I was able to build the native image.

**Here comes my first question**: We already included the package `org.hibernate` in `--initialize-at-build-time` so, why is it necessary to add that specific class? I though that adding the package everything is "included".

Then if I start the application and try to access the database it fails with:
```
Caused by: java.lang.IllegalArgumentException: Invalid logger interface org.hibernate.internal.CoreMessageLogger (implementation not found in java.lang.ClassLoader@5606aff25af8)
	at org.jboss.logging.Logger$1.run(Logger.java:2556)
	at java.security.AccessController.doPrivileged(AccessController.java:63)
	at org.jboss.logging.Logger.getMessageLogger(Logger.java:2529)
	at org.jboss.logging.Logger.getMessageLogger(Logger.java:2516)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.<clinit>(HqlSqlBaseWalker.java:38)
	at com.oracle.svm.core.hub.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:350)
	at com.oracle.svm.core.hub.ClassInitializationInfo.initialize(ClassInitializationInfo.java:270)
```

I've also tried adding `org.hibernate.internal.CoreMessageLogger,org.hibernate.internal.CoreMessageLogger_$logger` to `--initialize-at-build-time` but I have the same error.

Any thoughts?

**Comment by cstancu** [comment]
@graemerocher the reported error has been fixed by https://github.com/oracle/graal/commit/c6237d219a91c82b1954dfbfa0898ed917db09eb which was merged after the RC19 release and as @ilopmar reports, it works with master. 

@vjovanov can you please weigh in on https://github.com/oracle/graal/issues/1295#issuecomment-493409318?

**Comment by graemerocher** [comment]
@vjovanov any thoughts?

**Comment by cstancu** [comment]
@graemerocher @ilopmar all the reported issues seem to have been fixed with Micronaut 1.2.0 and GraalVM 19.2.0. You just need `--initialize-at-run-time=io.netty.buffer.AbstractReferenceCountedByteBuf,io.netty.util.AbstractReferenceCounted` to properly config Netty. This will be merged upstream by https://github.com/netty/netty/pull/9515.

@graemerocher thanks for the easy to reproduce build! Here is the diff with all the required changes:
```
diff --git a/Dockerfile b/Dockerfile
index 19c9513..fd1acbd 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -1,7 +1,8 @@
-FROM oracle/graalvm-ce:1.0.0-rc16 as graalvm
+FROM oracle/graalvm-ce:19.2.0 as graalvm
 WORKDIR /work
+RUN gu install native-image
 COPY ./target/micronaut-petclinic-*.jar .
-RUN native-image --no-server -cp micronaut-petclinic-*.jar
+RUN native-image --initialize-at-run-time=io.netty.buffer.AbstractReferenceCountedByteBuf,io.netty.util.AbstractReferenceCounted --no-server -cp micronaut-petclinic-*.jar
 
 FROM frolvlad/alpine-glibc
 EXPOSE 8080
diff --git a/pom.xml b/pom.xml
index bb76b32..d0308ae 100644
--- a/pom.xml
+++ b/pom.xml
@@ -8,7 +8,7 @@
   <version>0.1.0.BUILD-SNAPSHOT</version>
 
   <properties>
-    <micronaut.version>1.1.2.BUILD-SNAPSHOT</micronaut.version>
+    <micronaut.version>1.2.0</micronaut.version>
     <exec.mainClass>com.example.micronaut.petclinic.PetClinicApplication</exec.mainClass>
 
     <java.version>1.8</java.version>
```
I played around with the app a little bit (adding owners, pets, searching, etc.) and didn't come accross any errors.

**Comment by ilopmar** [comment]
@cstancu This is great, thank you very much!
